core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/promisehandlers/TakeGlobalCheckpointHandler.scala

```diff
@@ -2,87 +2,98 @@ package edu.uci.ics.amber.engine.architecture.controller.promisehandlers
-          if (!estimationOnly) {
-            // serialize CP state
-            chkpt.save(SerializedState.CP_STATE_KEY, this.cp)
-            logger.info(
-              s"Serialized CP state, current workflow state = ${cp.workflowExecution.getState}"
-            )
-            // get all output messages from cp.transferService
-            chkpt.save(
-              SerializedState.OUTPUT_MSG_KEY,
-              this.cp.transferService.getAllUnAckedMessages.toArray
-            )
-            val storage = SequentialRecordStorage.getStorage[CheckpointState](Some(uri))
-            val writer = storage.getWriter(actorId.name)
-            writer.writeRecord(chkpt)
-            writer.flush()
-            writer.close()
+    controllerInterface
+      .propagateChannelMarker(
+        PropagateChannelMarkerRequest(
+          cp.workflowExecution.getAllRegionExecutions
+            .flatMap(_.getAllOperatorExecutions.map(_._1))
+            .toSeq,
+          msg.checkpointId,
+          NO_ALIGNMENT,
+          cp.workflowScheduler.physicalPlan.operators.map(_.id).toSeq,
+          physicalOpIdsToTakeCheckpoint.toSeq,
+          PrepareCheckpointRequest(msg.checkpointId, estimationOnly),
+          METHOD_PREPARE_CHECKPOINT.getBareMethodName
+        ),
+        mkContext(SELF)
+      )
+      .flatMap { ret =>
+        Future
+          .collect(ret.returns.map {
+            case (workerId, _) =>
+              val destActor = ActorVirtualIdentity(workerId)
+              workerInterface
+                .finalizeCheckpoint(
+                  FinalizeCheckpointRequest(msg.checkpointId, uri.toString),
+                  mkContext(destActor)
+                )
+                .onSuccess { resp =>
+                  totalSize += resp.size
+                }
+                .onFailure { err =>
+                  throw err // TODO: handle failures.
+                }
+          }.toSeq)
+          .map { _ =>
+            logger.info("Start to take checkpoint")
+            val chkpt = new CheckpointState()
+            if (!estimationOnly) {
+              // serialize CP state
+              chkpt.save(SerializedState.CP_STATE_KEY, this.cp)
+              logger.info(
+                s"Serialized CP state, current workflow state = ${cp.workflowExecution.getState}"
+              )
+              // get all output messages from cp.transferService
+              chkpt.save(
+                SerializedState.OUTPUT_MSG_KEY,
+                this.cp.transferService.getAllUnAckedMessages.toArray
+              )
+              val storage = SequentialRecordStorage.getStorage[CheckpointState](Some(uri))
+              val writer = storage.getWriter(actorId.name)
+              writer.writeRecord(chkpt)
+              writer.flush()
+              writer.close()
+            }
+            totalSize += chkpt.size()
+            logger.info(s"global checkpoint finalized, total size = $totalSize")
+            TakeGlobalCheckpointResponse(totalSize)
```

Comment:
isn't this a standard IntResponse?

or LongResponse?

Reply:
Although it's just a number for now, I may need to add more information in this response, let's keep it for now.